### PR TITLE
Call pattern formatting

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary

This PR adds changes to delegate the formatting for match pattern nodes in the `MatchCase` formatting i.e., making the `pattern.format` calls. The implementation goes on to preserving the leading/trailing comments on the pattern nodes, parenthesizing if there are any such comments. If the node itself is parenthesized without any comments, the parentheses are preserved as is done in Black.

## Test Plan

Added test cases for:
* Pattern with parentheses, without leading or trailing comments
* Pattern with parentheses, with trailing comments
* Pattern with parentheses, with leading comments
* Pattern with parentheses, with leading and trailing comments

### CODE REVIEW

File: crates/ruff_python_formatter/src/other/match_case.rs
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/other/match_case.rs:23]: This now fails for cases where the pattern has comments. We may need to preserve parentheses similar to the `WithItem` formatting.

Comment (dhruvmanila) [reply crates/ruff_python_formatter/src/other/match_case.rs]: @MichaReiser This is fixed

File: crates/ruff_python_formatter/src/other/match_case.rs
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/other/match_case.rs:28]: Nit
```suggestion
        if comments.has_leading_comments(pattern) {
            pattern.format().fmt(f)?;
        } else {
```

Do we need to handle trailing own line comments too? I guess my question is. Does this implement the parenthesizing exactly like black or is the goal to unblock the work on match cases and we follow up with a matching formatting later?

**Comment by MichaReiser** [comment]
Thank you

File: crates/ruff_python_formatter/src/other/match_case.rs
Comment (dhruvmanila) [reply crates/ruff_python_formatter/src/other/match_case.rs:28]: By trailing, do you mean the following?

```python
match foo:
    case (
        # test
    foo  # trailing 1
    # trailing 2
# trailing 3
    ):
        pass
```

If so, they're handled correctly:

```python
match foo:
    case (
        # test
        x as NOT_YET_IMPLEMENTED_PatternMatchAs  # trailing 1
        # trailing 2
        # trailing 3
    ):
        pass
```

Comment (dhruvmanila) [reply crates/ruff_python_formatter/src/other/match_case.rs:28]: Oh, also this change isn't necessary as we need the leading comments to be added 😅

**Comment by MichaReiser** [comment]
Lol, I can't approve the change

**Comment by charliermarsh** [comment]
I'm so confused by this lol. Did Micha start this then Dhruv took over?

**Review by dhruvmanila — APPROVED** [review]
😉

File: crates/ruff_python_formatter/src/other/match_case.rs
Comment (dhruvmanila) [reply crates/ruff_python_formatter/src/other/match_case.rs:28]: Ah, you're correct, without the leading comments the parentheses aren't added.

**Comment by dhruvmanila** [comment]
@MichaReiser I went ahead and implemented the "only trailing comments" and "no comments" part as well. I've added test cases and the output matches `black`.

> I'm so confused by this lol. Did Micha start this then Dhruv took over?

Me and Micha discussed on Discord regarding calling the `pattern.format()` so that we can unlock https://github.com/astral-sh/ruff/issues/5834#issuecomment-1678848297 but that involved handling comments around the pattern. The implementation itself is similar to that of `StmtWith` formatting.

**Comment by MichaReiser** [comment]
@dhruvmanila feel free to merge this when it's ready

**Pull request merged**
